### DISCUSSION TITLE
Custom section commands (listing)?

### DISCUSSION THREAD

**Discussion by davidchisnall** [Omikhleia/resilient.sile#108] [Q&A] [answered]
The package includes captions for tables and figures.  I also need to be able to display code listings.  I assumed that the sectioning support would be able to make them.  I've created the following package that I *thought* created the right styles:

```lua
local base = require("packages.resilient.base")
local package = pl.class(base)
package._name = "listings"

function package:_init(options)
	base._init(self)

	self:registerStyle("listing-caption-base-number", {}, {})

	self:registerStyle("listing-caption-main-number", { inherit = "listing-caption-base-number" }, {
		numbering = { before = { text = "Listing " }, after = { text = ".", kern = "iwsp" } },
		font = { features = "+smcp" },
	})
	self:registerStyle("listing-caption-ref-number", { inherit = "listing-caption-base-number" }, {
		numbering = { before = { text = "listing " } },
	})

	self:registerStyle("listing-caption", {}, {
		font = { size = "4em" },
		paragraph = {
			before = { indent = false, vbreak = false },
			align = "center",
			after = { skip = "medskip" },
		},
		sectioning = {
			counter = { id = "listing", level = 2 },
			settings = {
				toclevel = 7,
				bookmark = false,
				goodbreak = false,
			},
			numberstyle = {
				main = "listing-caption-main-number",
				reference = "listing-caption-ref-number",
			},
		},
	})
	self:registerCommand("listingcaption", function(options, content)
		options.style = "listing-caption"
		SILE.call("sectioning", options, content)
	end, "Listings caption")
end

return package
```

And the following simple test:

```sil
\begin[class="resilient.book"]{document}
\use[module=packages.listings]{}
\ref[marker=lst:listing]{}
\ref[marker=lst:listing2]{}
\ref[marker=lst:listing3]{}
\sectioning[marker=lst:listing, style=listing-caption]{Here is a caption}
\sectioning[numbering=true, style=table-caption]{\label[marker=lst:listing2]{}Here is another caption}
\listingcaption[marker=lst:listing3]{Here is a third caption}
\sectioning[marker=fig, style=figure-caption]{Here is a figure third caption}
\end{document}
```

The `figure-caption` and `table-caption` ones are simply to check whether I understand how `\sectioning` works.  It seems to do what I expect.  But the captions for the two that use the `listing-caption` are prefixed with 'Table' (small caps), and not 'Listing'.  I'm really confused about where that's coming from.  Looking at the resilient.book class, I thought the `numbering = { before = { text = "` bit was defining this, but modifying my installed copy of `sile/classes/resilient/book.lua` to put something else in `table-caption-main-number` style doesn't change it both the `listing-caption` and `table-caption` lines are prefixed with Table.  

The `figure-caption` line is correctly prefixed with 'Figure', so obviously the infrastructure is working correctly and I am doing something wrong, but I'm not sure what.

**Comment by Omikhleia** [comment]
I'm getting this with your code:

![image](https://github.com/user-attachments/assets/33e83b6c-f041-4ddd-93aa-edf8bfa11e72)

Isn't it what you expected?

**Reply by Omikhleia** [reply]
BTW. Note that as per the documentation, once the file has been processed and its style file generated, the Lua definitions used as default aren't used any longer - So also check you don't have a style file for your document with some prior definitions (and if so, remove the offending styles so they'll be re-generated)

**Reply by Omikhleia** [reply]
Another detail (for now), but you might want to move your style registrations in a dedicated package method:

```
function package:registerStyles ()
   self:registerStyle("listing-caption-base-number", {}, {})
   ...
end
```

I's not important in the current implementation, but its probably more future proof.

**Reply by davidchisnall** [reply]
> Note that as per the documentation, once the file has been processed and its style file generated, the Lua definitions used as default aren't used any longer - So also check you don't have a style file for your document with some prior definitions (and if so, remove the offending styles so they'll be re-generated)

Aha, that was it.  I missed that in the documentation.  I was reading the `sectioning` docs and there wasn't a cross-reference to the documentation for styles so I was just reading the code (I couldn't find where `registerStyle` is implemented, so I was guessing based on uses).  I had tried deleting all of the `test.*` files except `test.sil`, but completely missed that styles don't follow the same naming convention as other cache files SILE generates.

**Reply by davidchisnall** [reply]
One other thing I'm confused about:

I don't see anything that uses the numberstyle.reference field.  I expected that to provide the style for the number in references, but it doesn't seem to.

**Reply by Omikhleia** [reply]
> I don't see anything that uses the numberstyle.reference field. I expected that to provide the style for the number in references, but it doesn't seem to.

Correct. The corresponding feature on the labelrefs.sile package wasn't implemented: https://github.com/Omikhleia/labelrefs.sile/issues/2

While it works "in most scenarios", labelrefs has some deep issues (some of which were discussed here https://github.com/Omikhleia/labelrefs.sile/issues/1) which are still present in "bare" SILE -- The "multiple package instanciation" question is still existing.

I still have it under my radar, in my tracklist for 2025, but I haven't taken any decision yet. There's a possibility I'll deprecate labelrefs.sile entirely, and bring it back entirely into resilient.sile, and then complete the planned styling for it. But there's so much to do in various areas that I won't commit to any deadline.

**Reply by davidchisnall** [reply]
Thanks. I can work around it quite easily (I already do for HTML output, where the numbers for cross references need to be computed separately anyway), I just didn't want to write code to work around the lack of a feature that actually exists.

**Reply by Omikhleia** [reply]
@davidchisnall Your listing styles and commands could be part of the next resilient release -- I hadn't such a thing 'cause I didn't need it initially, but there's no reason otherwise. Do you mind me importing the code from this discussion? (also adding a `\listofliistings` convenience command on the way).

**Reply by Omikhleia** [reply]
(With due credits of course)

**Reply by davidchisnall** [reply]
Most of it is copied and pasted from your code with trivial changes, so no objections at all. It would be great to have in the default release. 

If you’re interested, the book I’m working on currently is [here](https://github.com/CHERIoT-Platform/book). This is the tooling that does the conversion from semantic markup to HTML or SILE is [here](https://github.com/davidchisnall/igk) (and is currently totally undocumented and not expected to be usable by anyone except me). The HTML version of the book is [here](https://cheriot.org/book/), I still need to make a few cleanups of the SILE version for the PDFs and generate indexes (I’m not happy with how the text in the doc comment blocks is rendered currently, I haven’t figured out how to disable indents for every paragraph in a range and the padding is not right.

**Reply by Omikhleia** [reply, marked as answer]
@davidchisnall A standard listing environment was added in [v2.6.0](https://github.com/Omikhleia/resilient.sile/releases/tag/v2.6.0) (#112)

I am of course interested in your feedback :cat: 
And if it meets your expectation, would you mark this question as "Answered"?

**Reply by davidchisnall** [reply]
Thanks, I’ll update and see if I can remove the custom one.
